### ISSUE TITLE
t.p.modules.PythonModule yields duplicate modules

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5030]
|<img alt="nkvoll's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| nkvoll reported|
|-|-|
|Trac ID|trac#5030|
|Type|defect|
|Created|2011-04-06 11:04:31Z|
|Branch|https://github.com/twisted/twisted/tree/t.p.modules-duplicates-5030|

There is a yield in twisted.python.modules._ModuleIteratorHelper that allows for duplicate modules being returned by iterModules when dealing with namespaced packages.

This results in i.e unit tests being run multiple times because Trials unit test loader does not ensure that the executed tests are unique.

See the attachments to [#5021](https://github.com/twisted/twisted/issues/5021) (reproduce.tar.gz and reproduce-multiple.tar.gz) for some examples that reproduce this error.

Attachments:

* [modules-namespace.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b79/b791c26dda0912d6ba7d288f17811402e3c4e5b7/8ef9c5d0ceb76f4cce3bd37be3e71cb166b0a811.patch) (618 bytes) - added by nkvoll on 2011-04-06 11:05:02Z - added the guard to the second yield
* [modules_namespace_test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b79/b791c26dda0912d6ba7d288f17811402e3c4e5b7/af3890e356179c7e122636ab74b5ec6e45db7719.patch) (2487 bytes) - added by nkvoll on 2011-04-06 14:27:32Z - test coverage
* [clean_up_entries.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b79/b791c26dda0912d6ba7d288f17811402e3c4e5b7/6bd9baa69cbe731dc099e8c88a477b7dc56036bf.patch) (367 bytes) - added by nkvoll on 2011-04-06 22:12:22Z - clean up the entries that are created
* [clean_up_sys_modules.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b79/b791c26dda0912d6ba7d288f17811402e3c4e5b7/5f6236b1ae046ff6b0eaaf25a327a6302bb4455f.patch) (1902 bytes) - added by nkvoll on 2011-04-06 23:29:05Z - clean up sys.modules after importing
* [5030-review-changes.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b79/b791c26dda0912d6ba7d288f17811402e3c4e5b7/86acb1d5a0c0356a890030cfa8e852db3932aab5.patch) (1939 bytes) - added by nkvoll on 2011-04-13 10:27:19Z - news fragment, test docstring, removing unnecessary code

<details><summary>Searchable metadata</summary>

```
trac-id__5030 5030
type__defect defect
reporter__nkvoll nkvoll
priority__normal normal
milestone__ 
branch__branches_t_p_modules_duplicates_5030 branches/t.p.modules-duplicates-5030
branch_author__nkvoll__mwh nkvoll, mwh
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1302087871000000 1302087871000000
changetime__1303763199000000 1303763199000000
version__None None
owner__exarkun exarkun
cc__mwh cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="nkvoll's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|nkvoll commented|
|-|-|

I added the same guard thats used for the other yield in the same function, and it seemingly fixes the issue (as far as I can tell).

I ran trial before the patch without any changes to the results.

This seems to be the underlying cause for [#5021](https://github.com/twisted/twisted/issues/5021), and fixing this solves that issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to nkvoll|
|-|-|

Thanks for the patch, nkvoll.  Since you also filed [[#5021](https://github.com/twisted/twisted/issues/5021)](https://github.com/twisted/twisted/issues/5021), I wonder if you can explain how this is different.  Otherwise, this ticket seems to just be a duplicate of that ticket.

Also, at a minimum, all changes require test coverage.  So to accept either this patch or the one on #5021, there will need to be unit tests which demonstrate the problem and demonstrate that the change fixes it.

**Comment by twisted-trac** [comment]
|<img alt="nkvoll's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|nkvoll commented|
|-|-|

This is the root cause of the issue I had when I created [[#5021](https://github.com/twisted/twisted/issues/5021)](https://github.com/twisted/twisted/issues/5021). Patching it at the lowest level seems to be the best solution, so I created this ticket instead. Should I mark #5021 as a duplicate of this?

I've created a unit test that fails when used with Twisted 11.0, and succeeds after the attached patch has been applied. Attaching it now.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_6"></a>|@mwhudson commented|
|-|-|

(In [31562]) Branching to 't.p.modules-duplicates-5030'

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_7"></a>|@mwhudson commented|
|-|-|

(In [31563]) apply the patch from the ticket, refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_8"></a>|@mwhudson commented|
|-|-|

(In [31564]) apply the test patch from the ticket, refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_9"></a>|@mwhudson commented|
|-|-|

Unfortunately, this new test breaks test_nonexistentPaths :/

I don't really understand how the tests are intended to work, I think probably most of them should be creating their own PythonPath objects and not implicitly using the global one by calling module level functions from t.p.modules.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_12"></a>|@mwhudson commented|
|-|-|

Well, the test passes again, so let's put this back in review.

**Comment by twisted-trac** [comment]
|<img alt="nkvoll's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|nkvoll commented|
|-|-|

In this case, PythonPath cannot be used as a substitute for using the global one because in order for the namespace packages to be resolved correctly, they have to be imported. PythonPath uses reflect.namedAny to load modules, which does not use the PythonPath and thus is unable to find the test_package.

So the unit test I added does the same as the other tests in the same module do: it temporarily replaces the contents of sys.path using the helper function defined in its base class that takes care of cleaning up after the test completes.

This only leaves sys.modules, which I'm adding a patch for (apply on the current branches/t.p.modules-duplicates-5030).

As you mentioned, there might be some incentive to update as many unit tests in that module as possible to use PythonPath if possible (which might already have been done, as I haven't spent enough time to understand the technicalities of _all_ the other unit tests in the module), but I think that is outside of the scope of this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun commented|
|-|-|

(In [31566]) Apply clean_up_sys_modules.patch

refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to nkvoll<br>@exarkun set status to `new`|
|-|-|

Thanks for the patch!

  1. `BasicTests` really should have had a docstring already.
  1. What's the motivation behind the new `addCleanup` call in `pathEntryWithOnePackage`?  The tests all seem to pass without it; is it necessary for the tests to be correct?
  1. `test_namespacedPackages` needs a docstring as well
  1. Hmmm.  We have some code in Twisted, maybe in more than one place, that makes creating directory trees easier than this.  It might be hard to reach, though.  Ah yes, `twisted.python.test.test_release.StructureAssertingMixin`.  It would be nice if this test (and maybe others in this module) could re-use that.  Can you file a ticket for moving that mixin somewhere public (which means adding unit tests and maybe some docs for it)?  Then eventually we can refactor this new test and make it a little nicer to read.
  1. In the change to `modules.py`, `not modname in yielded` reads a little oddly and isn't consistent with the earlier filtering check in this code.  This is a really minor point, but it would be nicer as `modname not in yielded`. :)
  1. A [ReviewProcess#Newsfiles news fragment] would be great as well :)

The fix itself seems correct to me.  The conditional inside the `for potentialTopLevel in children:` loop checks to see if a child directory entry being encountered represents a Python module (because it's a .py file, for example).  If it is, and it is not a duplicate, it is yielded.  However, if it is not, the other branch of the conditional checks to see if it might be a _package_ rather than just a module.  But no deduplication check was applied there, presumably because you can't have duplicate packages without namespace packages (I think).

Thanks again.  If you can address the points above, I think we can resolve this ticket.  The patch has been committed to a branch now, so if you can make future patches against that branch, that would be great!

**Comment by twisted-trac** [comment]
|<img alt="nkvoll's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|nkvoll commented|
|-|-|

To address your questions one by one:

 1. I'm not sure what such a docstring should include, as I'm not too familiar with the rest of the unit tests in that class.
 1. The motivation _was_ that the `test_nonexistentPaths` test failed as it found more modules than expected. Initially I thought that this was because the temporary paths were not cleaned up, but it was because the test I added left sys.modules in an unclean state. Since I added the try-finally to clean up sys.modules, this is not needed, and I'll remove the extra code.
 1. Done
 1. Done -> [#5050](https://github.com/twisted/twisted/issues/5050)
 1. Done
 1. Done

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra removed owner|
|-|-|

adding review keyword for last patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

(In [31671]) Apply 5030-review-changes.patch 

refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

(In [31672]) and news fragment

refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_21"></a>|@exarkun set owner to @exarkun|
|-|-|

This looks pretty good now.  I'm going to make a couple more minor tweaks:

  1. Rather than saying "Verify ..." in the test method docstring, just say what's being verified (all tests are supposed to verify something :)
  1. Methods should be separated from each other by two blank lines

and then merge the branch.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [31673]) Remove "Verify ..." boilerplate from test method docstring; insert a little bit of vertical whitespace

refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

Hmm.  There is a problem on Python 2.5 on Fedora, which appears to be missing the pkgutil module.  I'm not yet sure why that might be.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [31722]) Make certain pkgutil is available for this test

refs [#5030](https://github.com/twisted/twisted/issues/5030)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun set status to `closed`|
|-|-|

(In [31724]) Merge t.p.modules-duplicates-5030

Author: nkvoll
Reviewer: mwh, exarkun
Fixes: [#5030](https://github.com/twisted/twisted/issues/5030)

Add support to `twisted.python.modules` for namespace packages.
